### PR TITLE
[PREVIEW] Feature service up/down alert

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Configure alerting for send letter service](https://tools.hmcts.net/jira/browse/RPE-689)

### Change description ###

As part of alert module, appinsights resource had to be extracted from webapp module. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: infrastructure/is_down_alert.tf [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module "send-letter-up-alert" {
+  source = "git@github.com:hmcts/cnp-module-metric-alert"
+  location = "${azurerm_application_insights.appinsights.location}"
+  app_insights_name = "${azurerm_application_insights.appinsights.name}"
+
+  alert_name = "Send Letter is DOWN - RPE"
+  alert_desc = "Triggers when send letter service looks like being down within a 15 minutes timeframe."
+  app_insights_query = "union (dependencies | where data == 'FtpFileUploaded' and success =~ 'true' | summarize ['name'] = 'uploaded', ['counter'] = sum(itemCount)), (requests | where name == 'POST SendLetterController/sendLetter' and success =~ 'true' | where tostring(operation_SyntheticSource) == '' | summarize ['name'] = 'received', ['counter'] = sum(itemCount)), (traces | extend logger = customDimensions.LoggerName | where logger == 'uk.gov.hmcts.reform.sendletter.tasks.UploadLettersTask' and message has 'FTP downtime' | summarize ['name'] = 'downtime', ['counter'] = sum(itemCount)) | evaluate pivot(name, sum(counter)) | project ['is_operational'] = iif(received > 0 and downtime > 0, true, iif(received > 0, uploaded > 0, true)) | where is_operational == false"
```
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply infrastructure/is_down_alert.tf]: would multiline string like here https://www.terraform.io/docs/configuration/variables.html#strings work?

[the remark this answers is not in the document]
Comment (doncem) [reply infrastructure/is_down_alert.tf]: `is_operational` stands for a question to the series conditions which satisfy the hypothesis whether service is operational or not / up or down. Just made up the column name.

File: infrastructure/is_down_alert.tf [hunk 2]
```diff
@@ -0,0 +1,30 @@
+module "send-letter-up-alert" {
+  source = "git@github.com:hmcts/cnp-module-metric-alert"
+  location = "${azurerm_application_insights.appinsights.location}"
+  app_insights_name = "${azurerm_application_insights.appinsights.name}"
+
+  alert_name = "Send Letter is DOWN - RPE"
+  alert_desc = "Triggers when send letter service looks like being down within a 15 minutes timeframe."
+  app_insights_query = <<EOF
```
Comment (timja) [inline infrastructure/is_down_alert.tf]: ❤️

File: infrastructure/appinsights.tf [hunk 3]
```diff
@@ -0,0 +1,10 @@
+resource "azurerm_application_insights" "appinsights" {
```
Comment (timja) [inline infrastructure/appinsights.tf]: how come this is here? its created by the webapp module unless you're using shared infra usually

Comment (doncem) [reply infrastructure/appinsights.tf]: is needed by alert. and resource is not loaded until webapp module is resolved. `depends_on` did not work in module

File: infrastructure/is_down_alert.tf [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ | where name == 'POST SendLetterController/sendLetter' and success =~ 'true'
+ | where tostring(operation_SyntheticSource) == ''
+ | summarize ['name'] = 'received', ['counter'] = sum(itemCount)), (traces
+ | extend logger = customDimensions.LoggerName
+ | where logger == 'uk.gov.hmcts.reform.sendletter.tasks.UploadLettersTask' and message has 'FTP downtime'
+ | summarize ['name'] = 'downtime', ['counter'] = sum(itemCount))
+ | evaluate pivot(name, sum(counter))
+ | project ['is_operational'] = iif(received > 0 and downtime > 0, true, iif(received > 0, uploaded > 0, true))
+ | where is_operational == false
+EOF
```
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply infrastructure/is_down_alert.tf]: perhaps something like this (if possible)
```
union
  (dependencies
    | where data == 'FtpFileUploaded' and success =~ 'true'
    | summarize ['name'] = 'uploaded', ['counter'] = sum(itemCount)
  ),
  (requests
    | where name == 'POST SendLetterController/sendLetter' and success =~ 'true'
    | where tostring(operation_SyntheticSource) == ''
    | summarize ['name'] = 'received', ['counter'] = sum(itemCount)
  ),
  (traces
    | extend logger = customDimensions.LoggerName
    | where logger == 'uk.gov.hmcts.reform.sendletter.tasks.UploadLettersTask' and message has 'FTP downtime'
    | summarize ['name'] = 'downtime', ['counter'] = sum(itemCount)
  )
| evaluate pivot(name, sum(counter))
| project ['is_operational'] = iif(received > 0 and downtime > 0, true, iif(received > 0, uploaded > 0, true))
| where is_operational == false
```

[the remark this answers is not in the document]
Comment (doncem) [reply infrastructure/is_down_alert.tf]: It is possible.

btw, I found that azure analytics has it's own right side of button aka context menu if you press on the query window. It has `Pretify Query` option and it _pretifies_ back to what it is here in PR atm 😉

**Pull request merged**